lib/rom/command.rb

```diff
@@ -20,36 +20,31 @@ module ROM
   #
   # @private
   class Command
+    extend Dry::Initializer::Mixin
     include Dry::Equalizer(:relation, :options)
     include Commands
     include Pipeline::Operator
 
     extend ClassMacros
     extend ClassInterface
 
-    include Options
-
     defines :adapter, :relation, :result, :input, :register_as, :restrictable
 
-    option :type, allow: [:create, :update, :delete]
-    option :source, reader: true
-    option :result, reader: true, allow: [:one, :many]
+    # @attr_reader [Relation] relation The command's relation
+    param :relation
+
+    CommandType = Dry::Types['strict.symbol'].enum(:create, :update, :delete)
+    Result = Dry::Types['strict.symbol'].enum(:one, :many)
+
+    option :type, type: CommandType, optional: true
+    option :source, reader: true, optional: true, default: -> c { c.relation }
+    option :result, reader: true, type: Result
     option :input, reader: true
-    option :curry_args, type: Array, reader: true, default: EMPTY_ARRAY
+    option :curry_args, reader: true, default: -> _ { EMPTY_ARRAY }
```

Comment:
Ahh, I see. But your solution requires deep analysis of whether a value is frozen to the guts. Shallow freezing isn't enought. This in turn require more dependencies (more than zero which means infinitely more), or more code and more overhead. Tbh defaults are overhead by themselves, but we love them ;)

That's why I'm strongly against this kind of interface simplification

(Actually I've made a same try in May, not via analyzis of freezing, but value dup-ing under the hood. That turns out to became a disaster, and I've cut and burn it eventually)